### PR TITLE
Implement API design with Postgres Backend

### PR DESCRIPTION
**Pull request by zacharyDez**
I implemented the first version of the API design with some small data processing scripts for ETL purposes in this PR:
- Add summary response with mocked data: Initial setup to return mocked data in the summary response.
- Add data ingestion: Implemented data ingestion to support data operations.
- Adapt API to read data: Modified the API to utilize the new data ingestion logic for enhanced data retrieval performance.
- Fix and clean up H3 generation functionality: Improve the H3 index generation logic for better accuracy and performance.
- Add tests on API: Introduced initial tests for API endpoints to ensure reliability and correctness.
- Adapt API for storing data within PostgreSQL: Transitioned to PostgreSQL for data storage, added functionality to load sample data for NYC, and created a visualization notebook with Lonboard for quick quality assurance.
- Add configuration variable for table name: Made the table name configurable through environment variables for flexibility.
- Add unit tests for db_utils and update existing API tests: Refactored API logic to use utility functions from db_utils.py, updated existing tests, and added new unit tests for database utility functions using pytest and unittest.mock.
- Fix definition of environment variables error: Resolved an issue with the order of environment variable definitions, ensuring they are loaded correctly. Added a notebook visualization that includes the available fields endpoint.

### CODE REVIEW

File: postgres/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+psycopg2-binary
+s3fs
```
Comment (vincentsarago) [inline postgres/requirements.txt:2]: why do we need those? 

it seems that in the `.py` files we mostly use `pandas` and `h3`

File: space2stats_api/app/routers/api.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    @field_validator("coordinates")
+    def validate_coordinates(cls, v):
+        try:
+            geom = shape({"type": "Polygon", "coordinates": v})
+            if not geom.is_valid:
+                raise ValueError("invalid coordinates for polygon")
+        except Exception as e:
+            raise ValueError("invalid coordinates for polygon") from e
+        return v
```
Comment (vincentsarago) [inline space2stats_api/app/routers/api.py:31]: we could maybe use https://github.com/developmentseed/geojson-pydantic

File: space2stats_api/app/routers/api.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+def get_summary(request: SummaryRequest):
+    try:
+        geom = shape({"type": "Polygon", "coordinates": request.aoi.coordinates})
+        if not geom.is_valid:
+            raise ValueError("invalid geojson polygon")
+    except Exception as e:
+        raise HTTPException(status_code=400, detail=str(e))
+
+    resolution = 6
+    aoi_geojson = mapping(geom)
```
Comment (vincentsarago) [inline space2stats_api/app/routers/api.py:52]: same if we use geojson-pydantic we might not need shapely functions

File: space2stats_api/app/routers/api.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        summary = {"hex_id": row[0], "fields": {col: row[idx] for idx, col in enumerate(colnames[1:], start=1) if col in request.fields}}
+        summaries.append(summary)
+
+    return summaries
+
+@router.get("/fields", response_model=List[str])
+def fields():
+    try:
+        fields = get_available_fields()
+        print(fields)
```
Comment (vincentsarago) [inline space2stats_api/app/routers/api.py:76]: remote `print`

File: space2stats_api/app/utils/db_utils.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+load_dotenv("../db.env")
+
+DB_HOST = os.getenv("DB_HOST")
+DB_PORT = os.getenv("DB_PORT")
+DB_NAME = os.getenv("DB_NAME")
+DB_USER = os.getenv("DB_USER")
+DB_PASSWORD = os.getenv("DB_PASSWORD")
+DB_TABLE_NAME = os.getenv("DB_TABLE_NAME")
+
+print(DB_TABLE_NAME)
```
Comment (vincentsarago) [inline space2stats_api/app/utils/db_utils.py:15]: we could use Pydantic-settings https://github.com/developmentseed/tipg/blob/0975f181f28aa94e11ccf3bf8737cd8734421018/tipg/settings.py#L116-L159

File: space2stats_api/app/utils/db_utils.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
+import os
+import psycopg2
```
Comment (vincentsarago) [inline space2stats_api/app/utils/db_utils.py:2]: IMO we should use psycopg 3 (most of the code here should be compatible)

File: space2stats_api/app/routers/api.py (hunk 3 above)
Comment (zacharyDez) [reply space2stats_api/app/routers/api.py]: So, we don't need them in `api.py`. I updated the code to use `geojson_pydantic` instead. However, Shapely is still used in the `h3_utils.py` for the spatial join function. We could off-load this logic to Postgres or use another method. I'm not sure if we would want to include those changes in this PR. @vincentsarago @bitner.

File: postgres/requirements.txt (hunk 1 above)
Comment (zacharyDez) [reply postgres/requirements.txt]: see https://github.com/worldbank/DECAT_Space2Stats/pull/8/files#diff-3bd785b4bbbd346532389bb929cd18816234cc5607ce7c5ac9dc207c2cc0a66a for the full list of requirements. There are some dependencies used in the utility functions called in `api.py`.

File: space2stats_api/app/utils/db_utils.py (hunk 5 above)
Comment (zacharyDez) [reply space2stats_api/app/utils/db_utils.py]: > we could use Pydantic-settings https://github.com/developmentseed/tipg/blob/0975f181f28aa94e11ccf3bf8737cd8734421018/tipg/settings.py#L116-L159

Implemented!

File: space2stats_api/app/routers/api.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    except Exception as e:
+        raise HTTPException(status_code=500, detail=str(e))
```
[the remark this answers is not in the document]
Comment (zacharyDez) [reply space2stats_api/app/routers/api.py]: Removed the exception, but running into some difficulties implementing the exception handler. I'll look into it when I have more time. Do you think it's a blocker to merge this PR?

[the remark this answers is not in the document]
Comment (vincentsarago) [reply space2stats_api/app/routers/api.py]: > Do you think it's a blocker to merge this PR?

No

File: space2stats_api/tests/test_api.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
+from fastapi.testclient import TestClient
```
Comment (vincentsarago) [inline space2stats_api/tests/test_api.py:1]: should we run the tests in the CI?

Comment (vincentsarago) [reply space2stats_api/tests/test_api.py:1]: yeah, you written nice tests they deserve to be ran in CI 😄

Comment (zacharyDez) [reply space2stats_api/tests/test_api.py:1]: First version implemented: https://github.com/worldbank/DECAT_Space2Stats/actions/runs/9999419959!

**Review by vincentsarago — APPROVED** [review]

**Pull request merged**